File: 912.py
from typing import List


class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        length = len(nums)
        if length < 2:
            return nums

        def merge(start, middle, end):
            l, r = start, middle + 1
            tmp = nums[start:middle+1]
            i = start
            l = 0
            # print(start, middle, end)
            while l <= middle - start and r <= end:
                # print(tmp[l])
                # print(nums[r])
                if tmp[l] < nums[r]:
                    nums[i] = tmp[l]
                    l += 1
                else:
                    nums[i] = nums[r]
                    r += 1
                i += 1
            if r > end:
                nums[i:end+1] = tmp[l:]

        def sort(start, end):
            if start == end:
                return
            middle = (start + end) // 2
            sort(start, middle)
            sort(middle + 1, end)
            merge(start, middle, end)

        sort(0, length - 1)
        return nums

    def sortArray1(self, nums: List[int]) -> List[int]:
        length = len(nums)
        if length < 2:
            return nums

        def merge(start, middle, end):
            l, r = start, middle + 1
            tmp = nums[start:middle+1]
            i = start
            l = 0
            # print(start, middle, end)
            while l <= middle - start and r <= end:
                # print(tmp[l])
                # print(nums[r])
                if tmp[l] < nums[r]:
                    nums[i] = tmp[l]
                    l += 1
                else:
                    nums[i] = nums[r]
                    r += 1
                i += 1
            if r > end:
                nums[i:end+1] = tmp[l:]

        count = 1
        while count < length:
            i = 0
            while i + count < length:
                merge(i, i + count - 1, min(i + 2 * count - 1, length - 1))
                i += 2 * count
            print(f'count:{count}')
            print(nums)    
            count <<= 1
        return nums


if __name__ == "__main__":
    s = Solution()
    print(s.sortArray1([4, 2, 3, 1]))

File: 912.go
package leetcode

import (
	"math/rand"
	"time"
)

func init() { rand.Seed(time.Now().UnixNano()) }

func sortArray(nums []int) []int {
	if len(nums) <= 1 {
		return nums
	}
	p := partion(nums, 0, len(nums))
	sortArray(nums[:p])
	sortArray(nums[p+1:])
	return nums
}

func partion(nums []int, low, high int) int {
	idx := rand.Int()%(high-low) + low
	nums[low], nums[idx] = nums[idx], nums[low]
	pivot := nums[low]
	i := low
	for j := low; j < high; j++ {
		if nums[j] <= pivot {
			nums[i], nums[j] = nums[j], nums[i]
			i++
		}
	}
	nums[low], nums[i-1] = nums[i-1], nums[low]
	// fmt.Println(nums, i-1)
	return i - 1
}

func sortArray2(nums []int) []int {
	n := len(nums)
	merge := func(i, size int) {
		if i+size >= n {
			return
		}
		tmp := append(make([]int, 0, size), nums[i:i+size]...)
		p1, p2 := 0, i+size
		r1, r2 := size, i+size*2
		for ; p1 < r1 && p2 < r2 && p2 < n; i++ {
			if tmp[p1] <= nums[p2] {
				nums[i] = tmp[p1]
				p1++
			} else {
				nums[i] = nums[p2]
				p2++
			}
		}
		for ; p1 < r1; p1++ {
			nums[i] = tmp[p1]
			i++
		}
	}

	size := 1
	for size < n {
		for i := 0; i < n; i += size * 2 {
			merge(i, size)
		}
		size <<= 1
	}
	return nums
}
